### PR TITLE
fix(stepfunctions-tasks): EvaluateExpression is limited to expressions that contain state paths

### LINKED ISSUE #7655: Why are tasks.EvaluateExpression limited to expressions with containing state paths
Perhaps there's a good reason to limit EvaluateExpression to those containing state paths?

### Use Case

If not, I was just trying to use it in a state machine to create a timestamp via `Date.now().toString()`

### Proposed Solution

Remove the constraint

### Other

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by nija-at**
### Commit Message
fix(stepfunctions-tasks): EvaluateExpression is limited to expressions that contain state paths (#7774)

fixes #7655
### End Commit Message

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by nija-at** [comment]
cc @jogold

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/evaluate-expression.ts [hunk 1]
```diff
@@ -56,18 +56,17 @@ export class EvaluateExpression implements sfn.IStepFunctionsTask {
   public bind(task: sfn.Task): sfn.StepFunctionsTaskConfig {
     const matches = this.props.expression.match(/\$[.\[][.a-zA-Z[\]0-9]+/g);
 
-    if (!matches) {
-      throw new Error('No paths found in expression');
+    let expressionAttributeValues: {} | undefined;
```
Comment (jogold) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/evaluate-expression.ts:59]: The handler does `Object.entries(event.expressionAttributeValues)` and will fail if `undefined`. I don't get why TypeScript doesn't get it here.

Comment (jogold) [reply packages/@aws-cdk/aws-stepfunctions-tasks/lib/evaluate-expression.ts:59]: OK the `as Event´ below is wrong that's why.

File: packages/@aws-cdk/aws-stepfunctions-tasks/lib/evaluate-expression.ts [hunk 2]
```diff
@@ -56,18 +56,17 @@ export class EvaluateExpression implements sfn.IStepFunctionsTask {
...
+    if (matches) {
+      expressionAttributeValues = matches.reduce(
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-stepfunctions-tasks/lib/evaluate-expression.ts:61]: This expression is a very complex, to me. Defining a helper function, doesn't it just read a lot better to say something like:

```ts
const matches = ...;
const expressionAttributeValues = fromEntries(matches.map(m => [m, sfn.Data.stringAt(m)]));
```
?

Obviously, before Node 12, this requires the helper:

```ts
function fromEntries<A>(entries: Array<[string, A]>): Record<string, A> {
  const ret: Record<string, A> = {};
  for (const [k, v] of entries)) {
    ret[k] = v;
  }
  return ret;
}
```

Not a blocker, I just generally prefer defining helper functions to make the main expression more readable...

**Comment by jogold** [comment]
As commented this will not work because `expressionAttributeValues` can now be `undefined`

**Comment by nija-at** [comment]
Thanks for the notice @jogold. I've updated the handler, and have tested this with an actual state machine. Let me know if this looks good to you.

**Comment by jogold** [comment]
> Thanks for the notice @jogold. I've updated the handler, and have tested this with an actual state machine. Let me know if this looks good to you.

I would personally have removed the `as Event` here, which is wrong because the typecast allows for an undefined `expressionAttributeValues`
https://github.com/aws/aws-cdk/blob/da6d40e7027c5f05b64c9ce591bbe96ca62f775f/packages/%40aws-cdk/aws-stepfunctions-tasks/lib/evaluate-expression.ts#L79-L82

This instead would have prevented the error in the first place:
```ts
    const parameters: Event = {
        expression: this.props.expression,
        expressionAttributeValues,
    }:
    return {
      resourceArn: evalFn.functionArn,
      policyStatements: [new iam.PolicyStatement({
        resources: [evalFn.functionArn],
        actions: ['lambda:InvokeFunction'],
      })],
      parameters,
    };
```

**Comment by nija-at** [comment]
Nice! That's definitely a better change. I've switched out the implementation to use your suggestion.

**Pull request merged**